pkg/curatedpackages/bundleclient.go

```diff
@@ -0,0 +1,172 @@
+	"github.com/aws/eks-anywhere-packages/pkg/bundle"
+	"github.com/aws/eks-anywhere/pkg/constants"
+)
+
+// BundleClient abstracts the source of package bundle information.
+//
+// Whether the source is a registry, a cluster, the BundleClient should
+// abstract the details of retrieving a particular curated packages bundle.
+type BundleClient interface {
+	ActiveOrLatest(ctx context.Context) (*packagesv1.PackageBundle, error)
+}
+
+// BundleClientOptions alter the behavior of the bundle client when retrieving
+// package bundle information.
+type BundleClientOptions struct {
+	// ClusterName identifies the cluster with which to communicate.
+	ClusterName string
+	// KubeConfig specifies a kubeconfig file to use.
+	//
+	// Defaults to the KUBECONFIG envrionment variable, and can fall back to
+	// ~/.kube/config or similar.
+	KubeConfig string
+	// KubeVersion specifies a kubernetes version requirement for package bundles.
+	//
+	// Required if consulting a registry for package bundle information.
+	KubeVersion string
+	// Registry as an OCI Registry URL.
+	//
+	// Defaults to the EKS-A ECR. Used when consulting a registry for package
+	// bundle information.
+	Registry string
+	// RegistryClient will be used if supplied.
+	//
+	// Specifying a custom registry client is optional, but could be needed
+	// when specifying a custom registry mirror that requires authentication.
+	RegistryClient bundle.RegistryClient
+	// restConfigurator is used to inject testing constructs.
+	//
+	// It shouldn't be required for normal use.
+	restConfigurator restConfigurator
+}
+
+// NewBundleClient constructs a BundleClient based on the value of source.
+func NewBundleClient(source BundleSource, opts BundleClientOptions) (BundleClient, error) {
+	switch source {
+	case Registry:
+		rc := opts.RegistryClient
+		if rc == nil {
+			rc = bundle.NewRegistryClient(artifacts.NewRegistryPuller())
+		}
+		kubeVersion := opts.KubeVersion
+		if kubeVersion == "" {
+			return nil, fmt.Errorf("kube version is required when source is registry")
+		}
+		registry := opts.Registry
+		if registry == "" {
+			registry = "public.ecr.aws/q0f6t3x4" // TODO pull this from the EKS-A bundle
+		}
+		return NewRegistryBundleClient(rc, registry, kubeVersion), nil
+	case Cluster:
+		kubeClient, err := newKubeClient(opts.KubeConfig, opts.restConfigurator)
+		if err != nil {
+			return nil, err
+		}
+		return NewClusterBundleClient(kubeClient, opts.ClusterName), nil
+	default:
+		return nil, fmt.Errorf("unhandled package bundle source %q", source)
+	}
+}
+
+// clusterBundleClient implements BundleClient by querying an existing cluster.
+type clusterBundleClient struct {
+	kubeClient  client.Client
+	clusterName string
+}
+
+var _ BundleClient = (*clusterBundleClient)(nil)
+
+// NewClusterBundleClient retrieves package bundle info from an existing cluster.
+func NewClusterBundleClient(kubeClient client.Client, clusterName string) *clusterBundleClient {
```

Comment:
Can we just make `clusterBundleClient`? or make the constructor private if it's not used from the outside

Reply:
No. If you could, I wouldn't have NewClusterBundleClient. :) The reason I say you can't, is because if you tried to instantiate your own clusterBundleClient, you wouldn't have a guarantee that you had kubeClient, or a clusterName, so every method would have to check for the existence of those things and throw a run-time error if they weren't defined. By making the struct not-exported, and the constructor exported, I let them be created only via the constructor, which requires the cluster name and the kubeClient.

I could make the constructor private I guess... But it doesn't buy me anything, since NewBundleClient calls it anyway, and that has to be exported... 

NewBundleClient is just a convenience method for lifting some common CLI code out of the separate CLI commands, and putting it here in this ... well, I guess it's factory method. But the point is there are several CLI commands that need to have a bundle client, for either a source argument, or a cluster argument, and NewBundleClient moves that repetitive logic out of the CLI commands to a single place that really knows more about it anyway.

Reply:
> I could make the constructor private I guess... But it doesn't buy me anything, since NewBundleClient calls it anyway, and that has to be exported...
There is a mayor difference: `NewBundleClient` return a public interface. So from there you don't return a private type, you return a public interface implemented by a private struct.

I just don't think it's a good idea to return private types from public methods in general. You leave the caller without information about what they can do with that return value or even what that value actually is. You would need to go a read the actual implementation of that type to know what public methods has and what they do.

Reply:
I'll make it private. If you get a struct that you don't know what it is, or how to use it, I consider that a strong hint that you shouldn't use that function is all. *shrug*